### DISCUSSION TITLE
`'/' is deprecated: Use a 'CasePathable' case key path, instead.`

### DISCUSSION THREAD

**Discussion by 3a4oT** [pointfreeco/swift-composable-architecture#3253] [Q&A]
Hello, I have a warning on a higher order reducer that says `'/' is deprecated: Use a 'CasePathable' case key path, instead.` 

```swift
@Reducer
public struct AnalyticsReducer<State, Action>: Sendable, Reducer where Action: AnalyticsTrackableAction {
    @Dependency(\.analyticsClient) private var analyticsClient
    public init() {}
    public func reduce(into state: inout State, action: Action) -> Effect<Action> {
        guard let event: AnalyticsEvent = (/Action.analyticsTrack).extract(from: action) else {
            return .none
        }
        return .run { _ in
            await analyticsClient.track(event)
        }
    }
}
```

How can I refactor `(/Action.analyticsTrack).extract(from: action)` to use  `CasePathable`?

Swift 5.10

**Comment by stephencelis** [comment]
The `AnalyticsTrackableAction` protocol needs to encode its ability to extract `analyticsTrack` actions.

The simplest would be to add a getter for the extraction:

```diff
 protocol AnalyticsTrackableAction {
   static func analyticsTrack(_ event: AnalyticsEvent) -> Self
+  var analyticsTrack: AnalyticsEvent? { get }
 }
```

Conforming enums would get this for free if you mark them with `@dynamicMemberLookup`:

```diff
+@dynamicMemberLookup
 @CasePathable enum Action: AnalyticsTrackableAction {
   case analyticsTrack(AnalyticsEvent)
 }
```

And then you would update your reducer operator:

```diff
-guard let event: AnalyticsEvent = (/Action.analyticsTrack).extract(from: action) else {
+guard let event: AnalyticsEvent = action.analyticsTrack else {
```

The alternative would be to constrain the protocol to `CasePathable` with an additional protocol constraint on `AllCasePaths`:

```swift
protocol AnalyticsTrackableAction: CasePathable
where AllCasePaths: HasAnalyticsEventCasePath<Self> {}

protocol HasAnalyticsEventCasePath<Root> {
  associatedtype Root: CasePathable
  var analyticsEvent: AnyCasePath<Root, AnalyticsEvent> { get }
}
```

And in practice you would need to extend both the action and its `AllCasePaths` type to the protocols for the conformance to work:

```swift
@CasePathable enum Action: AnalyticsTrackableAction {
  case analyticsEvent(Bool)
}
extension Action.AllCasePaths: HasAnalyticsEventCasePath {}
```

Note that in Swift <6 this extension needs to live in a separate file due to a bug with macros that was fixed in Swift 6.

**Reply by 3a4oT** [reply]
> The simplest would be to add a getter for the extraction:

Unfortunately, it doesn't work. I marked all related enums with `@dynamicMemberLookup` and when I extended the protocol with

```diff
 protocol AnalyticsTrackableAction {
   static func analyticsTrack(_ event: AnalyticsEvent) -> Self
+  var analyticsTrack: AnalyticsEvent? { get }
 }
```

I got compilation errors:

```
Type 'SomeReducer.Action' does not conform to protocol 'AnalyticsTrackableAction'

Candidate is an enum case with associated values, but protocol does not allow it
```

**Reply by stephencelis** [reply]
@3a4oT This code compiles just fine for me:

```swift

struct AnalyticsEvent {}

protocol AnalyticsTrackableAction {
  static func analyticsTrack(_ event: AnalyticsEvent) -> Self
  var analyticsTrack: AnalyticsEvent? { get }
}

@CasePathable
@dynamicMemberLookup
enum Action {
  case analyticsTrack(AnalyticsEvent)
}
```

**Reply by 3a4oT** [reply]
Thanks for the reply,  but should I conform **Action** to `AnalyticsTrackableAction`? Because when I do it:

```diff
@CasePathable
@dynamicMemberLookup
+ enum Action: AnalyticsTrackableAction {
  case analyticsTrack(AnalyticsEvent)
}
```
 The error is `Type 'Action' does not conform to protocol 'AnalyticsTrackableAction'` which is the same in my case

**Reply by stephencelis** [reply]
@3a4oT That was my bad! I was under the impression this worked and that we'd taken advantage of it elsewhere, but I'm mistaken. I think that unfortunately the other protocol-based solution is probably better, then, short of defining your own macro,

**Reply by pyrtsa** [reply]
Did you try adding the conformance in an extension instead? Possibly in a different file even?

**Reply by 3a4oT** [reply]
> Did you try adding the conformance in an extension instead? Possibly in a different file even?

`Enum 'case' is not allowed outside of an enum` - so you can't add the conformance in an extension.




> That was my bad! I was under the impression this worked and that we'd taken advantage of it elsewhere, but I'm mistaken. I think that unfortunately the other protocol-based solution is probably better, then, short of defining your own macro,

To be honest, I'm trying to avoid macros as much as possible in our codebase. I understand why TCA is using them, but in general, the macro system in Swift brings more problems than benefits at this point. Hopefully, it will be fixed in the future. I have one warning and around 80 places where I need to do a workaround with protocols and different files. I'll wait. Maybe TCA tools will come up with some smart ideas later on.

**Reply by lukeredpath** [reply]
@3a4oT I understand your concerns about macros but honestly, I think TCA is at a point now where trying to use it without macros is very very difficult, if not impossible. You can avoid certain conveniences like the `@Reducer` and `@DependencyClient` macros but so much stuff is driven by case key paths that trying to avoid `@CasePathable` is just going to result in a huge amount of manual code you'll need to implement.

**Reply by 3a4oT** [reply]
I'll rephrase. I'm OK with using macros from vendors like TCA (they're actively maintained and have good support). just don't want to maintain it internally and dance around `swift-syntax` versions or other things.

**Reply by stephencelis** [reply]
@3a4oT Did the second option work? It requires an extra line for the conformance but I _think_ I had it building locally.

In general we are open to ideas on how to better support generic case key paths (_i.e._ protocols that declare a particular case key path in a conformance), but we're not sure how to make it simpler than the above right now.

Maybe an enhancement to `@CasePathable` or another macro in the library would help, but we're not sure what it would look like.

**Reply by 3a4oT** [reply]
> ```swift
> extension SomeReducer.Action.AllCasePaths: HasAnalyticsEventCasePath {}
> ```

Unfortunately, it doesn't work either. I get a compilation error: `Type 'SomeReducer.Action.AllCasePaths' does not conform to protocol 'HasAnalyticsEventCasePath'`


I was trying on `Xcode 16 beta 4`
